### ISSUE TITLE
Arbitrary Data Format for HDFDataset

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#331]
It seems to me that the `HDFDataset` does only support data streams with a single time-axis, and not data with no or multiple time axes. I know that I can still add data without time axis, by just using a sequence length of one, and then remove the time axis in a layer in the network.

Are there any possibilities right now to load data with 2 time axes (e.g. Images) or also 1 time axis and 1 fixed spatial dimension into RETURNN? And if not, are there any plans yet to change that?

**Comment by albertz** [comment]
Yes, I do that a couple of times, and there are also some demos showing how to do that. You can simply flatten multiple dynamic axes into a single axis. When you dump to HDF (e.g. via `HDFDumpLayer`), this will be done automatically, and it will also save the information needed to unflatten it later. Then there is the `UnflattenNdLayer`, which unflattens the input. Search the tests for `unflatten_nd` to see some demos.

For fixed spatial dimensions, we do that very often, in a similar way. In that case, we merge it with the feature dim. E.g. when Sprint adds a window. Then you can simply use `SplitDimsLayer`.

**Comment by JackTemaki** [comment]
Okay. I would still leave this Issue open, as it would be nice to support more data-formats in the datasets itself in the future, without having to rely on extra layers.

**Comment by albertz** [comment]
I agree, we should simplify it for the user as far as possible. `HDFDumpLayer` and `HDFDataset` should complement each other in this case. (We should be a bit careful to not break old setups, though, when we change the behavior now.)

Another use case which came up a couple of times is to store a search beam in the HDF, e.g. N target hypotheses for one sequence. E.g. [this StackOverflow question](https://stackoverflow.com/questions/63411653/loading-ground-truth-and-hypotheses-as-training-target) (@npwynands @michelwi).

**Comment by albertz** [comment]
For reference, [here](https://github.com/rwth-i6/returnn/discussions/434) is some discussion relevant for this.

**Comment by albertz** [comment]
A note: I think our `Dataset` API itself is not really flexible enough.

The `Dataset.get_data_shape` function currently implies that there is always exactly one dynamic time axis in the beginning. So this is not only about `HDFDataset` but also about `Dataset` itself.

To make this really more generic, and support any number of dynamic axes by the dataset, we also would need sth equivalent to `DimensionTag` (but maybe simpler), to identify which of the dynamic time axis are the same and which are different. (Note that this is tricky to get right for all our existing datasets, because we never store this information explicitly. This is always implicitly known. You only notice in framewise training that you get strange errors when it doesn't fit...)

**Comment by JackTemaki** [comment]
I want to revive this issue, because we again ran into some trouble with the default HDF Dataset (writing sparse data with the `SimpleHDFWriter`). After reviewing some of the code, I think it would be best to have a closer look at how the `NextGenHDFDataset` is handling things, as it allows for a much cleaner way to create HDF files. While it certainly has some issues (e.g. the current parser naming), I think the concept of having such parsers is nice, as you can have custom parsers in your config. This would allow e.g. to store data with arbitrary compression algorithms inside the hdf and have a custom decompression logic.

Also, it has a cleaner and more consistent handling of multiple streams, compared to the unintuitive data vs targets split in the normal HDF.

Maybe one issue could be speed, as the NextGenHDF stores every sequence as its own dataset, which is maybe not a good idea in terms of performance, but I will do some tests with that.

I definitely do not want to extend/fix the code around the regular HDFDataset, as the codebase for the writer is already unnecessarily large and complex. Maybe even a clean re-write of any HDF logic would be best.

**Comment by albertz** [comment]
Despite the data / targets split (which is only a minor annoyance), I don't really see much problems with our current format.

But what you describe just sounds more like a minor bug in `SimpleHDFWriter`? Certainly writing sparse data is very much supported.

Also you say that our code is unnecessarily large and complex. Why don't we fix that? Inventing things over and over, writing new code again and again only that the new code also becomes complex then is not really such a good solution.

Instead of inventing yet another own custom format, and having a dataset implementation for that, I would suggest that we maybe use some of the existing very common formats. E.g. what HuggingFace uses for almost all datasets. It is based on Apache Arrow. See [here](https://huggingface.co/docs/datasets/about_arrow).

In your particular case, I would suggest to just fix the problem in `SimpleHDFWriter`. This sounds just like a small bug which is probably easy to fix.

**Comment by JackTemaki** [comment]
> Why don't we fix that? 

Because our policy of keeping full backwards compatibility prohibits that in many cases, and the personal capacity is not there. You know I am always in favor of cleaning up. But lets not discuss the problem of refactoring, this tackles some very fundamental issues that are not suited to be discussed here.

> Inventing things over and over
> Instead of inventing yet another own custom format

This is not what I planned here, the code already exists within RETURNN. The question is if it is a good idea to use it an if it needs some updates.

> Instead of inventing yet another own custom format, and having a dataset implementation for that, I would suggest [...]  existing very common formats [...] E.g. Apache Arrow

Now you are contradicting yourself. HDF is not an own custom format, and also integrating Apache Arrow would mean having new code and basically a rewrite of all the stuff we have (also in the recipes).

> This sounds just like a small bug which is probably easy to fix.

Maybe yes, but I do not like SimpleHDFWriter in the first place. It was never intended to be used outside of RETURNN, and I still think it was a bad decision of me to base some recipe jobs on it.

**Comment by albertz** [comment]
I prefer using an existing widely-used dataset format (HuggingFace dataset was just a suggestion or example for that) over `NextGenHDFDataset` or over inventing yet another own custom format.

I do not say that I prefer to use HuggingFace datasets over our `HDFDataset`. At least not for now. We first need some more experience with that.

I also do not say that we necessarily should use the HuggingFace dataset format. I just say I prefer a widely-used dataset format. That's just the only example I currently know.

Btw, what you read is probably on just the HuggingFace dataset API/Hub (all what you find [here](https://huggingface.co/datasets)) but I mean specifically the HuggingFace dataset format based on Apache Arrow (or  Apache Parquet?), which they use for many of the datasets. In most cases, they provide some scripts which load the dataset and that gets converted into their Apache Arrow based format, which is fully standardized and high performance. There is `save_to_disk`/`load_from_disk` (Apache Arrow based), `to_parquet`/`from_parquet` (Apache Parquet based), as far as I remember (I had some discussion on Librispeech here: https://github.com/huggingface/datasets/issues/4185).

The actual HuggingFace dataset format does also not matter too much, as they provide many ways how you can convert any data into their format. So you would probably use one of the intended ways how to do that. E.g. like almost all of their datasets are implemented. See for example Librispeech.

**Comment by JackTemaki** [comment]
> The actual HuggingFace dataset format does also not matter too much, as they provide many ways how you can convert any data into their format. So you would probably use one of the intended ways how to do that. E.g. like almost all of their datasets are implemented. See for example Librispeech.

This is not what we need the HDFDataset for. This is needed in our pipeline for converting arbitrary data, e.g. from RASR caches. Of course we could write a custom loader for the Huggingface dataset to load RASR caches. Or we could write a loader for HDF files with the huggingface dataset. But we are talking here about binary dataset formats. So HDF, npy/npz, protobuf etc.

And HDF is a standard, we have experience with that, we have existing scripts for that. We want to transfer single files, so multi file datasets are not good. So now it is only about the internal structuring which should be fully customizable. It should be possible to use HDFs with as few assumptions as possible. 

So given your statement our option is to keep using the HDFDataset or spend some work on integrating huggingface with so far no experience or even knowledge about binary formats, which potentially takes weeks to finish when no one puts priority on that. Neither is an option I want to work with.

I guess for short term convenience we will use the NextGenHDFDataset as is, and revise this topic completely when a more general direction is clear.

**Comment by albertz** [comment]
I'm not really sure I understand what you mean. We would just have a generic `HuggingFaceDataset` RETURNN dataset. This is https://github.com/rwth-i6/returnn/issues/1257. And this is very simple. Actually, we already have it (via @dthulke), it's just not pushed to master yet. This is for reading HuggingFace datasets. This includes both their binary formats (Apache Arrow, Apache Parquet, whatever else they have, maybe even HDF) and also any other of their datasets. It does not take weeks. We already have that. It's very simple. Maybe speak with @dthulke on how simple this is. Also, we probably will anyway have this, totally independent of the issue here, as it is very convenient to use all the existing HuggingFace datasets from their hub.

Apache Arrow, Apache Parquet is just as standard as HDF. ZIP files are also pretty widely used. The internal structure in all cases is fully customizable, so it's important to have a well defined format inside these container formats. Like the HuggingFace dataset format. Or our `HDFDataset`.

HDF can be a single file or multiple files. I think for Apache Arrow or Apache Parquet, this is just the same. It's probably just a matter of configuration. (In HF datasets, I think this is an option of `save_to_disk`.)

But anyway, I'm not stopping you from using and extending `NextGenHDFDataset`. I just think that this will overall lead us to more maintenance work than any of the other options.

**Comment by JackTemaki** [comment]
> This is for reading HuggingFace datasets

This is not the problem here. I do not doubt the integration is simple, but this does not solve the problem that we have the need for a file-based transfer format.

Please see the following links.
https://huggingface.co/docs/datasets/loading#local-and-remote-files
https://huggingface.co/docs/datasets/about_arrow#memorymapping
https://github.com/huggingface/datasets/issues/3113

All of this hints to that Huggingface datasets do NOT solve our problem.
Arrow is only used for local caching and not even listed as input type.

**Comment by albertz** [comment]
I still don't exactly understand what problem you are actually talking about (despite some small bug in `SimpleHDFWriter`). You ask about how to convert any of our data into a HuggingFace dataset? But that's also what I already answered: You would simply use one of the provided tools. Or you would simply write a HuggingFace dataset wrapper around your data, and then use functions like `save_to_disk` or `to_parquet`. We could also even write a HuggingFace dataset wrapper which wraps any RETURNN dataset. In any case, getting data into the HuggingFace dataset format should be just as easy.

So then you have your file-based transfer format. You just transfer whatever this `save_to_disk` or `to_parquet` (or maybe other tools or functions) give you. Where is the problem?

**Comment by albertz** [comment]
I again want to revive this issue here, and make it a bit more constructive. Let's start with some examples of data which is not directly supported.

- Scalars. It will always assume a time axis.
- One static dim, no time dim. We often would put a fixed-size vector in it and then assume the size is always the same.
- Multiple dynamic (time) dims.
- Etc

How can we adapt the current code that this can be supported?

Whatever data there is, it can always be flattened, so it can always be stored in the current HDF format. So I think what's only missing is then to store any of the additional needed meta info to restore the original data, e.g. restore multiple dyn dims, etc. For multiple dyn dims, it would need to store the seq lens then separately. And then, when loading, it can recover the original format.

Somewhat related to some of the issues is that `Dataset` currently still uses the `num_outputs` dict to describe its format, along with some functions like `get_dtype`, `get_shape`, etc., but it doesn't fully support everything that `Tensor`/`TensorDict` can represent. See #1302. This is basically required to properly support arbitrary data, because the `Dataset` API in any case then would need to be extended to be able to represent the data, and that is `Tensor`/`TensorDict`/#1302.

So I think we first need to resolve #1302, to discuss how to extend the `Dataset` API such that the API supports arbitrary data (independent of `HDFDataset`).

(cc @NeoLegends)

---

Of course, as it was mentioned, there are somewhat simple ways to work around those limitations. E.g. you can flatten multiple dynamic dims, also store the seq lens, and later unflatten them. That's how I often do it when I store data with multiple dyn dims. However, that's annoying, and doesn't mean we should not fix this issue.

---

And there were also discussions on using other kind of dataformats / some new dataset implementation (e.g. based on Apache Arrow, Parquet, HuggingFace datasets, or maybe reviving `NextGenHDFDataset`, or whatever else). Those discussions are also valid. But I think those are separate discussions, and we should discuss any more specific suggestions in separate issues. I'm very open to also add some new kind of dataset. Note, for that, again to support arbitrary data in it, again we need some way to properly represent/describe the data format, i.e. `Tensor`/`TensorDict`/#1302.
